ocaml-lsp-server/src/dune.ml

```diff
@@ -1,43 +1,251 @@
 open! Import
 
-type state =
+module Chan : sig
+  type t
+
+  val create : string -> t Fiber.t
+
+  val write : t -> Csexp.t list option -> unit Fiber.t
+
+  val read : t -> Csexp.t option Fiber.t
+
+  val pid : t -> Pid.t
+
+  val run : t -> unit Fiber.t
+end = struct
+  open Fiber.O
+
+  type t =
+    { in_thread : Scheduler.thread
+    ; out_thread : Scheduler.thread
+    ; out_chan : out_channel
+    ; in_chan : in_channel
+    ; pid : Pid.t
+    }
+
+  let pid t = t.pid
+
+  let write t sexp =
+    let+ res =
+      Scheduler.async_exn t.out_thread (fun () ->
+          match sexp with
+          | None ->
+            close_in_noerr t.in_chan;
+            close_out_noerr t.out_chan
+          | Some sexps ->
+            List.iter sexps ~f:(Csexp.to_channel t.out_chan);
+            flush t.out_chan)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let read t =
+    let+ res =
+      Scheduler.async_exn t.in_thread (fun () ->
+          match Csexp.input_opt t.in_chan with
+          | Error _
+          | Ok None ->
+            Unix.close (Unix.descr_of_in_channel t.in_chan);
+            None
+          | Ok (Some s) -> Some s)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let create dune =
+    let args = Array.of_list [ dune; "rpc"; "init"; "--wait" ] in
+    let pid, stdout, stdin =
+      let stdin_i, stdin_o = Unix.pipe () in
+      let stdout_i, stdout_o = Unix.pipe () in
+      let pid =
+        Unix.create_process dune args stdin_i stdout_o Unix.stderr |> Pid.of_int
+      in
+      (pid, stdout_i, stdin_o)
+    in
+    let* in_thread = Scheduler.create_thread () in
+    let+ out_thread = Scheduler.create_thread () in
+    let in_chan = Unix.in_channel_of_descr stdout in
+    let out_chan = Unix.out_channel_of_descr stdin in
+    { pid; in_chan; out_chan; in_thread; out_thread }
+
+  let run t =
+    let* (_ : Unix.process_status) = Scheduler.wait_for_process t.pid in
+    let+ () = write t None in
+    ()
+end
+
+module Client =
+  Drpc.Client
+    (struct
+      include Fiber
+
+      let parallel_iter t ~f =
+        let stream = Fiber.Stream.In.create t in
+        Fiber.Stream.In.parallel_iter stream ~f
+    end)
+    (Chan)
+
+type run =
   | Binary_not_found
   | Out_of_date
-  | Running
 
-type t = { state : state Fiber.Ivar.t }
+type state =
+  | Waiting_for_init of { diagnostics : Diagnostics.t }
```

Comment:
How can you have diagnostics if you're still waiting for init? If this is from a previous connection, wouldn't Closed need that as well?

Reply:
It's perhaps a bit confusing but `Diagnostics.t` is the thing that updates diagnostics on the client in a way that doesn't that does not remove any previous diagnostics from other sources (like merlin). When waiting for initialization, we need a reference to this value because it will be used when switching to the active state. In the closed state, we will no longer be able to transition to Acitive so the reference to diagnostics isn't needed anymore.